src/pages/workspaces/workspace/notebooks/NotebookLauncher.js

```diff
@@ -1,40 +1,43 @@
 import _ from 'lodash/fp'
-import { createRef, forwardRef, Fragment, useState } from 'react'
+import { forwardRef, Fragment, useRef, useState } from 'react'
 import { div, h, iframe } from 'react-hyperscript-helpers'
 import * as breadcrumbs from 'src/components/breadcrumbs'
 import { NewClusterModal } from 'src/components/ClusterManager'
-import { buttonOutline, linkButton, spinnerOverlay } from 'src/components/common'
-import { centeredSpinner, icon, spinner } from 'src/components/icons'
+import { buttonOutline, linkButton } from 'src/components/common'
+import { icon, spinner } from 'src/components/icons'
 import { notify } from 'src/components/Notifications'
-import { ajaxCaller } from 'src/libs/ajax'
-import colors from 'src/libs/colors'
-import { reportError } from 'src/libs/error'
+import { Ajax, useCancellation } from 'src/libs/ajax'
+import { withErrorReporting } from 'src/libs/error'
 import * as Nav from 'src/libs/nav'
 import * as Utils from 'src/libs/utils'
-import { Component } from 'src/libs/wrapped-components'
 import ExportNotebookModal from 'src/pages/workspaces/workspace/notebooks/ExportNotebookModal'
 import { wrapWorkspace } from 'src/pages/workspaces/workspace/WorkspaceContainer'
 
 
+const StatusMessage = ({ showSpinner, children }) => {
+  return div({ style: { padding: '1.5rem 2rem', display: 'flex' } }, [
+    showSpinner && spinner({ style: { marginRight: '0.5rem' } }),
+    div([children])
+  ])
+}
+
 const NotebookLauncher = _.flow(
   forwardRef,
   wrapWorkspace({
     breadcrumbs: props => breadcrumbs.commonPaths.workspaceTab(props, 'notebooks'),
     title: _.get('notebookName'),
     showTabBar: false
-  }),
-  ajaxCaller
-)(({ queryParams = {}, ...props }, ref) => {
-  const { workspace: { accessLevel, canCompute } } = props
+  })
+)(({ queryParams = {}, notebookName, workspace, workspace: { accessLevel, canCompute }, cluster, refreshClusters }, ref) => {
   return (Utils.canWrite(accessLevel) && canCompute && !queryParams['read-only']) ?
```

Comment:
TOL: This leads to one odd transition where, if you open the notebook read-only and see the preview then click "edit in Jupyter" while you don't have a cluster running, the read-only preview gets re-rendered. In both `NotebookEditor` (when in preview mode) and the alternate `Fragment` below, there's a message pane followed by the read-only notebook preview. Maybe capturing this condition and pushing it into `NotebookEditor` as an intent-to-edit gesture would avoid the regeneration of the preview.